File: patterns/multi-agent/code/typescript/vercel-ai-sdk/multi-agent.ts
/**
 * Multi-Agent — Vercel AI SDK variant.
 *
 * Pattern: Supervisor agent receives a task, decides which sub-agents to
 *   invoke and in what order, collects outputs, synthesizes the final result.
 * Framework: Vercel AI SDK (ai ^4.0.0) + @ai-sdk/anthropic (^1.0.0).
 * Idioms: generateObject() with a Zod Delegation schema owns the supervisor's
 *   "which agents do I call next" decision; each sub-agent is a plain
 *   `(task, context) => Promise<string>` function (could itself wrap generateText).
 *   No tools / no agent loop on the supervisor — the structured-output mode is
 *   the orchestration primitive.
 * Design doc: ../../../design.md (the framework-agnostic
 *   ../../python/multi_agent.py runs the same researcher → writer → reviewer
 *   delegation against the same enterprise-overview task).
 *
 * Install:  pnpm add ai @ai-sdk/anthropic zod
 *           (or via tsx-on-demand: npx tsx multi-agent.ts)
 * Run:      ANTHROPIC_API_KEY=... npx tsx multi-agent.ts
 *
 * Note: ESM only.
 * tsconfig snippet: { "module": "ESNext", "moduleResolution": "Bundler", "target": "ES2022" }
 */

import { anthropic } from "@ai-sdk/anthropic";
import { generateObject, generateText } from "ai";
import { z } from "zod";

// ── Types ────────────────────────────────────────────────────────────────────

interface SubAgent {
  name: string;
  description: string;
  run: (task: string, context: string) => Promise<string>;
}

interface DelegationLog {
  agentName: string;
  task: string;
  output: string;
}

interface MultiAgentResult {
  finalOutput: string;
  delegations: DelegationLog[];
  rounds: number;
}

// ── Supervisor schema ───────────────────────────────────────────────────────

const Delegation = z.object({
  done: z.boolean().describe("True when the supervisor judges the task is complete."),
  reason: z.string().optional(),
  next: z
    .array(z.object({
      agent: z.string().describe("Sub-agent name from the catalog."),
      task: z.string().describe("Sub-task description for that agent."),
    }))
    .describe("Sub-agents to invoke this round. Empty when `done` is true."),
});

function supervisorPrompt(agents: SubAgent[]): string {
  const catalog = agents.map((a) => `- ${a.name}: ${a.description}`).join("\n");
  return (
    "You are the supervisor of a multi-agent system. Each round, decide " +
    "whether the task is complete; if not, list the sub-agents to invoke " +
    "and what each should do. Return a JSON object with `done`, `reason`, " +
    `and \`next\`.\n\nAvailable agents:\n${catalog}`
  );
}

// ── System ──────────────────────────────────────────────────────────────────

interface MultiAgentOptions {
  agents: SubAgent[];
  maxRounds?: number;
}

class MultiAgentSystem {
  private readonly byName: Map<string, SubAgent>;

  constructor(
    private readonly opts: MultiAgentOptions,
    private readonly model = anthropic("claude-haiku-4-5"),
  ) {
    this.byName = new Map(opts.agents.map((a) => [a.name, a]));
    this.opts.maxRounds ??= 4;
  }

  async run(task: string): Promise<MultiAgentResult> {
    const delegations: DelegationLog[] = [];
    let context = "";

    for (let round = 1; round <= (this.opts.maxRounds ?? 4); round++) {
      const decision = await generateObject({
        model: this.model,
        schema: Delegation,
        system: supervisorPrompt(this.opts.agents),
        prompt: `Task: ${task}\n\nDelegations so far:\n${
          delegations.map((d) => `[${d.agentName}] ${d.task} -> ${d.output.slice(0, 80)}`).join("\n") || "(none)"
        }\n\nDecide next.`,
      });

      if (decision.object.done) {
        const final = await generateText({
          model: this.model,
          system: "You synthesize the final answer from the sub-agents' outputs.",
          prompt: `Task: ${task}\n\nSub-agent outputs:\n${delegations.map((d) => `[${d.agentName}] ${d.output}`).join("\n\n")}`,
        });
        return { finalOutput: final.text, delegations, rounds: round };
      }

      for (const call of decision.object.next) {
        const agent = this.byName.get(call.agent);
        if (!agent) continue;
        const output = await agent.run(call.task, context);
        delegations.push({ agentName: agent.name, task: call.task, output });
        context = `${context}\n\n[${agent.name}] ${output}`.trim();
      }
    }

    return {
      finalOutput: "Reached max rounds without supervisor signaling done.",
      delegations,
      rounds: this.opts.maxRounds ?? 4,
    };
  }
}

// ── Smoke runner ────────────────────────────────────────────────────────────

const makeAgentFn = (name: string) => async (task: string, ctx: string): Promise<string> =>
  `[${name} output] Task: ${task.slice(0, 50)} | Context: ${ctx.slice(0, 30) || "none"}`;

async function main(): Promise<void> {
  if (!process.env.ANTHROPIC_API_KEY) {
    console.log("Skipping smoke run — set ANTHROPIC_API_KEY to exercise the supervisor decision loop.");
    return;
  }

  const system = new MultiAgentSystem({
    agents: [
      { name: "researcher", description: "Finds and summarizes factual information from sources", run: makeAgentFn("researcher") },
      { name: "writer", description: "Writes clear, structured content based on provided research", run: makeAgentFn("writer") },
      { name: "reviewer", description: "Reviews content for accuracy, clarity, and completeness", run: makeAgentFn("reviewer") },
    ],
    maxRounds: 4,
  });

  const result = await system.run("Write a technical overview of LLM agent frameworks for a developer audience");
  console.log(`Delegations: ${result.delegations.length}`);
  for (const d of result.delegations) {
    console.log(`  -> [${d.agentName}] ${d.task.slice(0, 60)}`);
  }
  console.log(`\nFinal output:\n${result.finalOutput.slice(0, 200)}`);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}

export { MultiAgentSystem, Delegation, makeAgentFn };
export type { SubAgent, MultiAgentOptions, MultiAgentResult, DelegationLog };

File: patterns/multi_agent/code/python/multi_agent.py
"""
Multi-Agent — Supervisor delegates to specialized sub-agents.

A supervisor agent receives a task, decides which sub-agents to invoke
and in what order, collects their outputs, and synthesizes the final result.
Sub-agents are themselves autonomous (each can use tools, have memory, etc.).

Design doc:  ../../design.md
Overview:    ../../overview.md
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Callable, Protocol


# ── Interface ─────────────────────────────────────────────────────────────────

class LLM(Protocol):
    def generate(self, messages: list[dict]) -> str: ...


# ── Core types ────────────────────────────────────────────────────────────────

@dataclass
class SubAgent:
    name: str
    description: str                  # What this agent specializes in
    run: Callable[[str, str], str]    # fn(task, context) -> output


@dataclass
class Delegation:
    agent_name: str
    task: str
    context: str = ""


@dataclass
class AgentOutput:
    agent_name: str
    task: str
    output: str


@dataclass
class MultiAgentResult:
    final_output: str
    delegations: list[Delegation] = field(default_factory=list)
    agent_outputs: list[AgentOutput] = field(default_factory=list)


# ── Prompts ───────────────────────────────────────────────────────────────────

DELEGATE_PROMPT = """\
You are a supervisor agent. Decompose the following task and delegate
sub-tasks to the appropriate specialized agents.

Available agents:
{agents}

Task: {task}

Context from completed work so far:
{context}

Respond with a JSON array of delegations, or an empty array [] if the task
is now complete. Each delegation:
{{"agent": "<agent_name>", "task": "<specific sub-task to delegate>"}}

If work is complete, instead respond with:
{{"done": true, "reason": "<why the task is complete>"}}"""

SYNTHESIZE_PROMPT = """\
Synthesize the following agent outputs into a final result for the original task.

Original task: {task}

Agent outputs:
{outputs}

Produce the complete, unified final output."""


# ── Supervisor ────────────────────────────────────────────────────────────────

class MultiAgentSystem:
    """
    A supervisor that orchestrates multiple specialized sub-agents.

    The supervisor iteratively:
    1. Decides which agents to delegate to next
    2. Runs delegated tasks (sequentially by default; extend for parallel)
    3. Feeds outputs back as context
    4. Synthesizes the final result when done
    """

    def __init__(
        self,
        supervisor: LLM,
        agents: list[SubAgent],
        max_rounds: int = 5,
        shared_state: dict | None = None,
    ):
        self.supervisor = supervisor
        self.agents: dict[str, SubAgent] = {a.name: a for a in agents}
        self.max_rounds = max_rounds
        self.shared_state: dict = shared_state or {}

    def _decide(self, task: str, context: str) -> list[Delegation] | None:
        """Returns list of delegations, or None if done."""
        agent_list = "\n".join(
            f"- {a.name}: {a.description}" for a in self.agents.values()
        )
        messages = [{"role": "user", "content": DELEGATE_PROMPT.format(
            agents=agent_list, task=task, context=context or "None"
        )}]
        raw = self.supervisor.generate(messages)
        try:
            data = json.loads(raw)
            if isinstance(data, dict) and data.get("done"):
                return None  # Supervisor signals completion
            if isinstance(data, list):
                return [Delegation(agent_name=d["agent"], task=d["task"]) for d in data]
        except (json.JSONDecodeError, KeyError):
            pass
        return None

    def _delegate(self, delegation: Delegation, context: str) -> AgentOutput:
        agent = self.agents.get(delegation.agent_name)
        if not agent:
            output = f"Agent '{delegation.agent_name}' not found."
        else:
            output = agent.run(delegation.task, context)

        # Write to shared state so other agents can read it
        self.shared_state[delegation.agent_name] = output
        return AgentOutput(
            agent_name=delegation.agent_name,
            task=delegation.task,
            output=output,
        )

    def _synthesize(self, task: str, outputs: list[AgentOutput]) -> str:
        formatted = "\n\n".join(
            f"[{o.agent_name}]\nTask: {o.task}\nOutput: {o.output}"
            for o in outputs
        )
        messages = [{"role": "user", "content": SYNTHESIZE_PROMPT.format(
            task=task, outputs=formatted
        )}]
        return self.supervisor.generate(messages)

    def run(self, task: str) -> MultiAgentResult:
        all_delegations: list[Delegation] = []
        all_outputs: list[AgentOutput] = []
        context = ""

        for _ in range(self.max_rounds):
            delegations = self._decide(task, context)

            if delegations is None:  # Supervisor says done
                break

            if not delegations:
                break

            for delegation in delegations:
                output = self._delegate(delegation, context)
                all_delegations.append(delegation)
                all_outputs.append(output)
                context += f"\n[{output.agent_name}]: {output.output}"

        final = self._synthesize(task, all_outputs)
        return MultiAgentResult(
            final_output=final,
            delegations=all_delegations,
            agent_outputs=all_outputs,
        )


# ── Example ───────────────────────────────────────────────────────────────────

if __name__ == "__main__":

    class MockLLM:
        def __init__(self):
            self._round = 0

        def generate(self, messages: list[dict]) -> str:
            self._round += 1
            content = messages[-1]["content"]
            if "Available agents" in content:
                if self._round <= 1:
                    return json.dumps([
                        {"agent": "researcher", "task": "Research current LLM agent frameworks"},
                        {"agent": "writer", "task": "Write a summary based on research"},
                    ])
                return json.dumps({"done": True, "reason": "All sub-tasks complete"})
            return f"[synthesized output from {len(messages)} messages]"

    def make_agent_fn(name: str) -> Callable[[str, str], str]:
        def fn(task: str, context: str) -> str:
            return f"[{name} output] Task: {task[:50]} | Context: {context[:30] or 'none'}"
        return fn

    system = MultiAgentSystem(
        supervisor=MockLLM(),
        agents=[
            SubAgent(
                name="researcher",
                description="Finds and summarizes factual information from sources",
                run=make_agent_fn("researcher"),
            ),
            SubAgent(
                name="writer",
                description="Writes clear, structured content based on provided research",
                run=make_agent_fn("writer"),
            ),
            SubAgent(
                name="reviewer",
                description="Reviews content for accuracy, clarity, and completeness",
                run=make_agent_fn("reviewer"),
            ),
        ],
        max_rounds=4,
    )

    result = system.run("Write a technical overview of LLM agent frameworks for a developer audience")
    print(f"Delegations: {len(result.delegations)}")
    for d in result.delegations:
        print(f"  → [{d.agent_name}] {d.task[:60]}")
    print(f"\nFinal output:\n{result.final_output}")
